python-for-android/dists/kolibri/src/main/java/org/learningequality/Kolibri/TaskworkerWorker.java

```diff
@@ -36,4 +38,46 @@ public static Data buildInputData (String workerArgument, boolean longRunning) {
         Log.v(TAG, "Request data: " + data.toString());
         return data;
     }
+
+    protected void cleanup() {
+        hideNotification();
+        mWorker = null;
+    }
+
+    @Override
+    public ForegroundInfo getForegroundInfo() {
+        NotificationRef ref;
+        // If we are running in the service, use the service notification ref
+        if (TaskworkerWorkerService.mService != null) {
+            ref = TaskworkerWorkerService.mService.getNotificationRef();
+        } else {
+            ref = getNotificationRef();
+            Log.w(TAG, "No service found, using worker notification for foreground");
+        }
+
+        NotificationBuilder builder = new NotificationBuilder(getApplicationContext(), ref);
+
+        return new ForegroundInfo(ref.getId(), builder.build());
+    }
+
+    public NotificationRef getNotificationRef() {
+        // Use worker request ID as notification tag
+        String tag = getId().toString();
+        return new NotificationRef(NotificationRef.REF_CHANNEL_DEFAULT, tag);
+    }
+
+    public static void updateProgress(String notificationTitle, String notificationText, int progress, int total) {
+        if (mWorker != null) {
+            // We could also update progress on the worker here, if we need info about it on
+            // the Android side
+            // @see setProgressAsync
```

Comment:
Yeah - I guess the only thing would be if we wanted to handle the progress updating of the notification via Java rather than doing send notification here.